Optional ScalarType support for native functions & JIT

tugrulates opened the pull request:
For #6593 and #9515

This completes the support for optional<ScalarType> in native, JIT and autograd.

Note: Mostly following the existing implementation for optional<Scalar> that was added in https://github.com/pytorch/pytorch/pull/12582.

This PR introduces a way to make functions accept an optional dtype and it will unblock #9515 by allowing the `dtype` param for type promotion interface:
```
func: name(inputs, *, ScalarType? dtype=None, Casting casting=same_kind)
```

An alternative approach could have been using `ScalarType::Undefined` for the same purpose but without optional, though it would have been a bit hacky.
```
func: name(inputs, *, ScalarType dtype=Undefined, Casting casting=same_kind)
```

Here's an example use of this in action: https://github.com/pytorch/pytorch/pull/15133/commits/971f69eac69101955ed90078b44dab975d37a4f7

There are already a bunch of native functions that were getting optional `dtype` through function overloading. https://github.com/pytorch/pytorch/pull/15133 is the attempt to migrate all of those. I will send those changes separately after this since some functions (e.g. sum) need quite a bit of change in the codebase. See the commits over there.

tugrulates committed:
Optional ScalarType

tugrulates commented on torch/csrc/jit/passes/to_batch.cpp:
```diff
@@ -473,6 +473,7 @@ void ToBatch::toBatch(Block* block, Block* res_block) {
     else if(n->kind().is_prim()){
       switch(n->kind()){
         case prim::Constant:
+        case prim::None:
```
This just tells to copy `None` values in JIT graph to batch JIT graph as is.

tugrulates commented on torch/csrc/utils/python_arg_parser.cpp:
```diff
@@ -27,7 +27,6 @@ static std::unordered_map<std::string, ParameterType> type_map = {
   {"Storage", ParameterType::STORAGE},
   {"PyObject*", ParameterType::PYOBJECT},
   {"ScalarType", ParameterType::SCALARTYPE},
-  {"optional<ScalarType>", ParameterType::SCALARTYPE},
```
This wasn't used and not needed.

gchanan commented on tools/autograd/gen_python_functions.py:
```diff
@@ -769,7 +769,7 @@ def get_py_formal_arg(arg):
         if arg.get('default') is not None:
             default = arg['default']
-            if default == 'nullptr' or default == 'nullopt' or default == '{}':
+            if default == 'nullptr' or default == 'nullopt' or default == 'c10::nullopt' or default == '{}':
```
why do you need c10::nullopt?  where does that come from?

tugrulates replied:
I remember JIT was generating that (register_aten_ops) last week. It doesn't seem to be needed any more, so removed this.

tugrulates replied:
Ah, this was still needed. It's added in https://github.com/pytorch/pytorch/blob/b6edd7bbb4d0f79065ad6dac99b1a50575170039/aten/src/ATen/native_parse.py#L26

I will either revert this or find a different fix in the next PR.

tugrulates committed:
Remove unused code

zou3519 reviewed (commented):
Is there an easy way to add a test?

I don't know enough about this part of the JIT to comment on the correctness but it looks very similar to the original PR that added optional type. Maybe @wanchaol or @zdevito can take a look here?

wanchaol reviewed (commented):
Thank for adding this! the PR LGTM. Can you add a test in JIT to test a op that takes Optional[ScalarType], pass a None into it and see if it works end to end?

zasdfgbnm commented:
Could we add test for this by creating a dummy aten ops on `native_functions.yaml` to use these features, and test it on `test_jit.py` by using these ops? Sounds ugly, but easy to implement.

tugrulates commented:
I'm about to convert `?` suffix in type system to optional<Type>. Will add a test for `optional<Scalar>` in test_jit soon here.

tugrulates commented:
Added another commit which converts all optional types except `Tensor?` from `T?` to `optional<T>`. cc @gchanan 

**Why leave Tensor out?** `Tensor?` accepts undefined Tensors. There's a bunch of TODOs in the codebase to clean this up. I added a few more TODOs into locations that needs cleanup too.

**Why not wait for `Tensor?` to be cleaned up?** More and more needs for optional types and it's becoming harder to clean in ? suffix format.

#12582 added optional support by following `Tensor?` syntax. This PR now changes the syntax to use `optional<>`. @wanchaol @zdevito was there any discussion around this around that time? Any specific reason the optional<> syntax can horribly fail?

In any case, I can still drop the last commit and use `ScalarType?` for the original need in this PR.

tugrulates commented:
Existing JIT tests that use `optional<Device>` and `optional<Scalar>` should cover this end to end. I updated the expect files wherever needed. PTAL

There is currently no op using `optional<ScalarType>`. I have a stack of conversion here:
https://github.com/tugrulates/pytorch/commits/optional-scalartype

I'm holding off on sending them before this PR is resolved. I plan to add a JIT test for `optional<ScalarType>` with the first op that uses it.

wanchaol commented:
> Added another commit which converts all optional types except `Tensor?` from `T?` to `optional<T>`. cc @gchanan
> 
> **Why leave Tensor out?** `Tensor?` accepts undefined Tensors. There's a bunch of TODOs in the codebase to clean this up. I added a few more TODOs into locations that needs cleanup too.
> 
> **Why not wait for `Tensor?` to be cleaned up?** More and more needs for optional types and it's becoming harder to clean in ? suffix format.
> 
> #12582 added optional support by following `Tensor?` syntax. This PR now changes the syntax to use `optional<>`. @wanchaol @zdevito was there any discussion around this around that time? Any specific reason the optional<> syntax can horribly fail?
> 
> In any case, I can still drop the last commit and use `ScalarType?` for the original need in this PR.

I don't see the real value of migrating from `?` to `optional<>` syntax in schema. `?` is from the schema side, it is just a short hand of the `optional<>` syntax is c++ side, they mean the same thing. It was originally designed to use `?` to denote the optional types (as the same with bunch of other typed languages), it is just that `Tensor?` have slightly different meaning because of old TH behaviors. it is also good to keep the schema simple and easier to understand. 

Also `Tensor?` will always accept undefined tensor and this might not going to have any changes, we will not have optional<Tensor> afaik. Keeping the same and one syntax for those kind of stuff will reduce the mental overhead as well.

tugrulates commented:
Reverted the last commit that changes the schema syntax to `optional<T>`. I will add the e2e tests with the next PR that migrates the ops to use this (https://github.com/pytorch/pytorch/pull/15133). This one should be ready to go.

wanchaol reviewed (approved):
Nice
